### PR TITLE
[SPARK-6797][SPARKR] Add support for YARN cluster mode.

### PR DESCRIPTION
**Pull request by sun-rui**
This PR enables SparkR to dynamically ship the SparkR binary package to the AM node in YARN cluster mode, thus it is no longer required that the SparkR package be installed on each worker node.

This PR uses the JDK jar tool to package the SparkR package, because jar is thought to be available on both Linux/Windows platforms where JDK has been installed.

This PR does not address the R worker involved in RDD API. Will address it in a separate JIRA issue.

This PR does not address SBT build. SparkR installation and packaging by SBT will be addressed in a separate JIRA issue.

R/install-dev.bat is not tested. @shivaram , Could you help to test it?

### CODE REVIEW

**Comment by rxin** [comment]
Why is this needed if the only thing it is using is the DataFrame API?

**Comment by rxin** [comment]
ah i see. this is only for the AM.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -373,6 +374,22 @@ object SparkSubmit {
...
+      if (sparkHome.isEmpty) {
+        printErrorAndExit("SPARK_HOME does not exist for R application in yarn mode.")
+      }
+      val rPackagePath = Seq(sparkHome.get, "R", "lib").mkString(File.separator)
+      val rPackageFile = new File(rPackagePath, SPARKR_PACKAGE_ARCHIVE)
+      if (!rPackageFile.exists()) {
+        printErrorAndExit(s"$SPARKR_PACKAGE_ARCHIVE does not exist for R application in yarn mode.")
+      }
+      val localURI = Utils.resolveURI(rPackageFile.getAbsolutePath)
+      args.archives = mergeFileLists(args.archives, localURI.toString + "#sparkr")
```
Comment (shivaram) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Could you add a comment here as to why we have this '#sparkr" ? I believe this is to get the archive to unzip to a symlink named `sparkr` ?

**Comment by sun-rui** [comment]
Yes this assigns a symbol  link name. Thus we can refer to the shipped package via the logical name instead of the specific archive file name.

**Comment by shivaram** [comment]
Thanks @sun-rui -- I didn't get a chance to test this on windows (and a YARN cluster yet). Will try to do it this weekend.

Also it looks like there is a merge conflict. Could you resolve that ?

**Comment by sun-rui** [comment]
rebased

File: R/install-dev.bat [hunk 2]
```diff
@@ -25,3 +25,8 @@ set SPARK_HOME=%~dp0..
 MKDIR %SPARK_HOME%\R\lib
 
 R.exe CMD INSTALL --library="%SPARK_HOME%\R\lib"  %SPARK_HOME%\R\pkg\
+
+REM Zip the SparkR package so that it can be distributed to worker nodes on YARN
+pushd %SPARK_HOME%\R\lib
+jar cfM "%SPARK_HOME%\R\lib\sparkr.zip" SparkR
```
Comment (shivaram) [inline R/install-dev.bat]: @sun-rui -- this should be `jar.exe` instead of `jar`. The other thing is that `jar.exe` is only available in the JDK and not in the JAR version. So sometimes this may not be in the `PATH`. There are a couple of options for things we can do here

1. We can use `%JAVA_HOME%\bin\jar.exe` -- This might be more safer as users need to set JAVA_HOME for the compilation to work correctly
2. Rtools [1] by default installs a zip utility [2] as `zip.exe`. At least on my machine running `zip.exe -r sparkr.zip SparkR` seems to work.

[1] http://cran.r-project.org/bin/windows/Rtools/
[2] http://www.info-zip.org/

File: R/install-dev.bat [hunk 3]
```diff
@@ -25,3 +25,8 @@ set SPARK_HOME=%~dp0..
 MKDIR %SPARK_HOME%\R\lib
 
 R.exe CMD INSTALL --library="%SPARK_HOME%\R\lib"  %SPARK_HOME%\R\pkg\
+
+REM Zip the SparkR package so that it can be distributed to worker nodes on YARN
```
Comment (shivaram) [inline R/install-dev.bat]: Lets use `rem` instead of `REM` to be consistent with the rest of the file

**Comment by shivaram** [comment]
@sun-rui Thanks for the update. I just tested this on a YARN cluster and things seem to work correctly for the use case where we create a data frame from a file (i.e read.df).

However the `createDataFrame` code path still requires the RRDD to pick up `worker.R` from the zip file. Is there a separate JIRA we have for that ?

cc @davies

**Comment by sun-rui** [comment]
I originally planned to have a separate JIRA issue for adding shipping of the SparkR package for RDD APIs. But if this is still required by DataFrame API, I can do it in this PR.
Just wonder to know why all test pass? Not enough test cases for createDataFrame()?

**Comment by shivaram** [comment]
The tests pass because the `createDataFrame` call only fails in the YARN cluster mode and we don't have tests for SparkR in the YarnSuite (Actually this deserves a new JIRA).

BTW the change needed for `createDataFrame` to work should be minimal -- I think we can just set the variable `SPARKR_PACKAGE_DIR` for all modes and then set the default value of `sparkRLibDir`[1] to be `Sys.getenv(SPARKR_PACKAGE_DIR)`.

[1] https://github.com/apache/spark/blob/master/R/pkg/R/sparkR.R#L103

**Comment by tgravescs** [comment]
I'm curious, have you looked at ways of shipping R itself with the job or are you relying on R being installed on all the yarn nodes?

**Comment by sun-rui** [comment]
@tgravescs， I think the problem of shipping R itself is that R executable is platform specific. Also it may require OS specific installation before running R (not sure). pySpark also does not ship python itself.

**Comment by sun-rui** [comment]
Add support for shipping SparkR package for R workers required by RDD APIs. Tested createDataFrame() by creating a DataFrame from an R list.

Remove sparkRLibDir  parameter of sparkR.init(). Determine SparkR package location on each worker node according to the deployment mode (this allows node-specific SPARK_HOME). Not sure if there is better solution. A rough code scan about pySpark does not tell me how pySpark locates pySpark.zip in various deployment modes. @davies , could you help to give me a hint and review this patch?

Next, I'd like to refactor this code to align with SPARK-5479 (moves YARN specific code from SparkSubmit to deploy/yarn) @shivaram, do you think I refactor code in this patch or do it in a new JIRA?

**Comment by sun-rui** [comment]
rebased

**Comment by shivaram** [comment]
Thanks @sun-rui for the update. @davies can probably confirm but AFAIK the PySpark location is picked up at [1] by looking at the JAR file path / Spark home.

I'll take a closer look at the code later today, but in terms of refactoring I'd say lets do in a separate JIRA.

[1] https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala#L31

**Comment by davies** [comment]
@sun-rui Unfortunately, I also don't know much about how PySpark run on YARN. Could you add some unit test for YARN mode? (just follow the Python ones)

**Comment by sun-rui** [comment]
@shivaram, yes, I saw that function, but felt confusing that it does not consider the YARN mode case. 
@davies, it seems unit tests for pySpark in YARN modes were added in https://github.com/apache/spark/pull/6360/files. Need confirmation.

**Comment by shivaram** [comment]
cc @andrewor14 who probably knows some thing about the YARN tests

**Comment by davies** [comment]
@sun-rui Yes

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.api.r
+
+import java.io.{File}
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: remove `{ }`

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.apache.spark.SparkException
+
+private[spark] object RUtils {
+  /**
+   * Get the SparkR package path in various deployment modes.
+   */
+  def sparkRPackagePath(driver: Boolean): String = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: I would just pass in an extra parameter `master: String` into this method. We can get that from `RRDD#context`

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    val yarnMode = sys.env.get("SPARK_YARN_MODE")
+    if (!yarnMode.isEmpty && yarnMode.get == "true" &&
+        !(driver && System.getProperty("spark.master") == "yarn-client")) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: this expression is really hard to read. What are you trying to say?

File: core/src/main/scala/org/apache/spark/deploy/RRunner.scala [hunk 7]
```diff
@@ -71,9 +71,10 @@ object RRunner {
         val builder = new ProcessBuilder(Seq(rCommand, rFileNormalized) ++ otherArgs)
         val env = builder.environment()
         env.put("EXISTING_SPARKR_BACKEND_PORT", sparkRBackendPort.toString)
-        val sparkHome = System.getenv("SPARK_HOME")
+        val rPackageDir = RUtils.sparkRPackagePath(true)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/RRunner.scala]: can you do `driver = true` here to be more explicit?

File: core/src/main/scala/org/apache/spark/api/r/RRDD.scala [hunk 8]
```diff
@@ -390,9 +386,10 @@ private[r] object RRDD {
     thread
   }
 
-  private def createRProcess(rLibDir: String, port: Int, script: String): BufferedStreamThread = {
+  private def createRProcess(port: Int, script: String): BufferedStreamThread = {
     val rCommand = "Rscript"
     val rOptions = "--vanilla"
+    val rLibDir = RUtils.sparkRPackagePath(false)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RRDD.scala]: same here

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      // The SparkR package distributed as an archive resource should be pointed to
+      // by a symbol link "sparkr" in the current directory.
+      new File("sparkr").getAbsolutePath
+    } else {
+      // TBD: add support for MESOS
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: IIUC, this case is for standalone mode and the driver on yarn-client mode. Is that correct?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 10]
```diff
@@ -339,6 +340,24 @@ object SparkSubmit {
...
     }
 
+    // In yarn mode for an R app, add the SparkR package archive to archives
+    // that can be distributed with the job
+    if (args.isR && clusterManager == YARN) {
+      val sparkHome = sys.env.get("SPARK_HOME")
+      if (sparkHome.isEmpty) {
+        printErrorAndExit("SPARK_HOME does not exist for R application in yarn mode.")
+      }
+      val rPackagePath = Seq(sparkHome.get, "R", "lib").mkString(File.separator)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Is there a way to make this also go through `RUtils.sparkRPackagePath`? It seems weird to have two separate code paths here.

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 4 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: fixed

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 5 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: This function can be called on both driver and worker side. RDD#context will throw an exception if called on worker side.

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 6 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: The intention is that:
driver at yarn cluster  mode -> get sparkR package from current directory
driver at yarn client mode -> get sparkR package from $SPARK_HOME/R/lib
worker at yarn cluster  mode -> get sparkR package from current directory
worker at yarn client mode -> get sparkR package from current directory

I can add comments to make it easy to be read.

File: core/src/main/scala/org/apache/spark/deploy/RRunner.scala (hunk 7 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/deploy/RRunner.scala]: done

File: core/src/main/scala/org/apache/spark/api/r/RRDD.scala (hunk 8 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRDD.scala]: done

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 9 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: correct:)

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 10 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: fixed.

**Comment by sun-rui** [comment]
rebased

**Comment by shivaram** [comment]
@sun-rui The reason this patch keeps failing the MiMa tests is that we are modifying the constructor of RRDD which is a Developer API if I am not wrong. Anyways that gives us a good moment to think if we want to remove `rLibDir` from `sparkR.init` as well. Is there some use case where users might want to change the SparkR install location ?

**Comment by sun-rui** [comment]
@shivaram, the rLibDir parameter of sparkR.init() was intended for locating SparkR package on worker nodes at the time when SparkR was a separate project. Since now SparkR is a part of Spark, I think this parameter is useless because:
1. For YARN modes, SparkR package is dynamically shipped to workers and located in the current working directory, uses do not need to care about its location. This patch is for this purpose;
2. For standalone mode, SparkR is part of the Spark distribution, and can be located within the specific sub-directory of the distribution. This patch allows worker specific SPARK_HOME, which means it is not required that SPARK_HOME across all workers be the same. While rLibDir parameter assumes all workers have the same location of SparkR ( a limitation)
3. For Meos mode, spark.mesos.executor.home is used to specify location of Spark on workers. SparkR can be located relative to this location.

This question makes me think about SparkConf.setSparkHome(), which confuses me during creation of this patch. I am not sure how is setting is honored across different deployment modes.

**Comment by sun-rui** [comment]
@davies, I will add test case for YARN modes in a separate PR which is intended to update SparkR related logic to align with SPARK-5479.

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    val yarnMode = sys.env.get("SPARK_YARN_MODE")
+    if (!yarnMode.isEmpty && yarnMode.get == "true" &&
+        !(driver && System.getProperty("spark.master") == "yarn-client")) {
+      // For workers in YARN modes and driver in yarn cluster mode,
+      // the SparkR package distributed as an archive resource should be pointed to
+      // by a symbol link "sparkr" in the current directory.
+      new File("sparkr").getAbsolutePath
+    } else {
+      // TBD: add support for MESOS
+      val rPackagePath = localSparkRPackagePath
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: doesn't seem like `localSparkRPackagePath` will ever return empty because of this `map` call?
```
sparkHome.map(
      Seq(_, "R", "lib").mkString(File.separator)
    )
```

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: it will be empty if `SPARK_HOME` is not set right? This is an option map.

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      if (rPackagePath.isEmpty) {
+        throw new SparkException("SPARK_HOME not set. Can't locate SparkR package.")
+      }
+      rPackagePath.get
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: you can just rewrite this else case as
```
localSparkRPackagePath.getOrElse {
  throw new SparkException("SPARK_HOME not set...")
}
```

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+   * Get the SparkR package path in various deployment modes.
+   */
+  def sparkRPackagePath(driver: Boolean): String = {
+    val yarnMode = sys.env.get("SPARK_YARN_MODE")
+    if (!yarnMode.isEmpty && yarnMode.get == "true" &&
+        !(driver && System.getProperty("spark.master") == "yarn-client")) {
+      // For workers in YARN modes and driver in yarn cluster mode,
+      // the SparkR package distributed as an archive resource should be pointed to
+      // by a symbol link "sparkr" in the current directory.
+      new File("sparkr").getAbsolutePath
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: I think we should rewrite this method in the following way. Right now it reads the deploy mode from the master, which happens to work for YARN but isn't really the right way to do it. Instead, we should pass the deploy mode from `SparkSubmit.scala` as `spark.submit.deployMode` or something. Then this method would look like
```
/**
 * Get the SparkR package path in various deployment modes.
 * This assumes that system properties `spark.master` and `spark.submit.deployMode`
 * and environment variable `SPARK_HOME` are set.
 */
def sparkRPackagePath(driver: Boolean): String = {
  val master = sys.props("spark.master")
  val deployMode = sys.props("spark.submit.deployMode")
  val isYarnCluster = master.contains("yarn") && deployMode == "cluster"
  val isYarnClient = master.contains("yarn") && deployMode == "client"

  // In YARN mode, the SparkR package is distributed as an archive symbolically
  // linked to the "sparkr" file in the current directory. Note that this does not apply
  // to the driver in client mode because it is run outside of the cluster.
  if (isYarnCluster || (isYarnClient && !driver)) {
    new File("sparkr").getAbsolutePath
  } else {
    // Otherwise, assume the package is local
    // TODO: support this for Mesos
    localSparkRPackagePath.getOrElse {
      throw new SparkException("SPARK_HOME not set. Can't locate SparkR package.")
    }
  }
}
```

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 6 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: even with comments it's still hard to follow. I suggested an alternative below.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 14]
```diff
@@ -339,6 +341,23 @@ object SparkSubmit {
...
+      }
+      val rPackageFile = new File(rPackagePath.get, SPARKR_PACKAGE_ARCHIVE)
+      if (!rPackageFile.exists()) {
+        printErrorAndExit(s"$SPARKR_PACKAGE_ARCHIVE does not exist for R application in yarn mode.")
+      }
+      val localURI = Utils.resolveURI(rPackageFile.getAbsolutePath)
+
+      // Assigns a symbol link name "sparkr" to the shipped package.
+      args.archives = mergeFileLists(args.archives, localURI.toString + "#sparkr")
+    }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: github won't let me comment down there, but in L388 of this file we should add the deploy mode to a system property:
```
OptionAssigner(args.deployMode, ALL_CLUSTER_MGRS, ALL_DEPLOY_MODES,
  sysProp = "spark.submit.deployMode")
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 15]
```diff
@@ -339,6 +341,23 @@ object SparkSubmit {
       }
     }
 
+    // In yarn mode for an R app, add the SparkR package archive to archives
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: nit: elsewhere we use capital YARN

**Comment by andrewor14** [comment]
@sun-rui this looks pretty close. I left a suggestion to rewrite a method for readability, but functionally it looks good. Have you had a chance to test this on a real YARN cluster, with or without additional R libraries?

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 11 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: @felixcheung, sys.env.get() returns Option[String], Option[T].map() applies to the Option's value if the Option is nonempty.

**Comment by sun-rui** [comment]
@andrewor14, 
I have tested this patch with a real YARN cluster. 

For an R program, it can source other R files and call functions within them, or it can call functions exposed in other R packages. For sourcing R files, it can be supported via the existing --files (refer to SPARK-6833). For dependent R packages, SparkR now only supports passing the name list of the R packages to workers, but not distributing the packages themselves. It is expected that they are already installed in the R environment on worker nodes. Not sure if we support distributing dependent R packages or not, @shivaram, what do you think?

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 12 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: done

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 15 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: done

**Comment by shivaram** [comment]
Not yet -- As a part of #7139 @brkyvz is adding support for installing Spark packages with R code from source. We can then extend that to install CRAN R packages as well in the future. That need not be a concern for this PR.

**Comment by shivaram** [comment]
BTW if we are changing the RRDD constructor we should add a MimaExcludes for it to make the tests pass. @andrewor14 or @davies can probably give more information about this

**Comment by andrewor14** [comment]
@sun-rui you could just copy and paste the error message in the test output to `MimaExcludes`. For instance, you should add the following line there:
```
ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.api.r.BaseRRDD.this")
```

**Comment by sun-rui** [comment]
@andrewor14, I did a modification on your proposed code because when sparkRPackagePath() is called on executors, Most of the Spark conf properties can't be accessed from the Java system properties because the executor backends do not update the system properties according to the sparkConf. Fortunately, there is a global SparkEnv object that holds a sparkConf in this case.

**Comment by sun-rui** [comment]
rebased

File: project/MimaExcludes.scala [hunk 16]
```diff
@@ -81,6 +81,10 @@ object MimaExcludes {
               "org.apache.spark.mllib.linalg.Matrix.numNonzeros"),
             ProblemFilters.exclude[MissingMethodProblem](
               "org.apache.spark.mllib.linalg.Matrix.numActives")
+          ) ++ Seq(
+            // SPARK-6797 Support YARN modes for SparkR
+            ProblemFilters.exclude[MissingMethodProblem](
+              "org.apache.spark.api.r.BaseRRDD.this")
```
Comment (andrewor14) [inline project/MimaExcludes.scala]: sorry I wasn't being clear. There are a few more things you need to add here. The one I pointed out is just an example.

**Comment by andrewor14** [comment]
@sun-rui OK not a problem. I think this is ready to go once it passes tests. LGTM!

**Comment by sun-rui** [comment]
The unit test failure:
[info] - Single command with -e *** FAILED *** (2 minutes, 18 seconds)
[info]   java.util.concurrent.TimeoutException: Futures timed out after [2 minutes]
[info]   at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
[info]   at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
[info]   at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)
[info]   at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
[info]   at scala.concurrent.Await$.result(package.scala:107)
[info]   at org.apache.spark.sql.hive.thriftserver.CliSuite.runCliWithin(CliSuite.scala:92)

Seems not related to my change.

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by andrewor14** [comment]
@sun-rui could you rebase this? There are conflicts now. @shivaram does the R side look good to you? If so feel free to merge it.

**Comment by shivaram** [comment]
Yeah R code looks fine to me. However as this changes some fundamental things with how we locate the package, @sun-rui it will be great if you could confirm the two or three scenarios work correctly (with just a local master or a standalone master is fine):
1. Using shell by running `bin/sparkR`
2. Running a script using `bin/spark-submit` or `bin/sparkR`
3. Launching SparkR from RStudio 

If things look good lets merge this once its rebased + tested. We can do more testing while its in the tree before 1.5

**Comment by sun-rui** [comment]
@shivaram , tests done. Also tested with YARN cluster, yarn-client, standalone, createDataFrame() in YARN client mode.

**Comment by shivaram** [comment]
Thanks @sun-rui - LGTM. Merging this

**Comment by shivaram** [comment]
@sun-rui Could you close this PR ? Looks like Github PRs are not being closed due to an infrastructure issue https://issues.apache.org/jira/browse/INFRA-9988

**Comment by sun-rui** [comment]
@shivaram, close the PR.
